src/main/java/cuchaz/enigma/SourceProvider.java

```diff
@@ -71,7 +72,8 @@ public CompilationUnit getSources(String name) {
 
 		AstBuilder builder = new AstBuilder(context);
 		builder.addType(resolvedType);
-		builder.runTransformations(null);
+		// todo fix insert conversion transform so it works on livingentity
+		builder.runTransformations(transform -> !(transform instanceof InsertNecessaryConversionsTransform));
```

Comment:
What does this transformer do? (is it important?)

Reply:
dunno, but it currently crashes the whole living entity class and threaded anvil chunk storage class.

Reply:
It adds unchecked casts. At first, Procyon adds casts present in the bytecode (`checkcast`). After generic type inference, it removes unnecessary casts (of values returned by generic function) through the `RemoveUnnecessaryConversions` transform, and also adds necessary casts which were not present in the bytecode (unchecked casts) through the `InsertNecessaryConversions` transform.

Removing it will cause a few more type errors, but also fix a few cases where casts are wrongly being added.

Thiakil's Procyon fork has fixes for this without removing the transformation.

My Procyon fork rewrites this and other parts of Procyon to fix almost all type errors, but it causes Enigma's transformations to fail on a few classes.

Reply:
Most of the time, our procyon decompilation fails on this transformer, so I figured cutting off on this one may be a smart choice.

Reply:
Yes, I agree that removing it is ok. It's broken, and anyway the code generated by Procyon is very far from being error-free without rewriting the type inference system.

Maybe you should also test with Thiakil's fork to see if you get even better results, and if yes, we should switch to it.

https://github.com/thiakil/procyon

Reply:
or we can just pull yours back and do some minor tweak around your type inference system. yours should be reliable for the most part

Reply:
Yes, but I'm currently busy with school and don't have much time to work on it for a while.

If you want to look at it, the issue is that two values of different types are sometimes determined by the inferer that they must have the same types. You can try taking a method on which it breaks, reducing it to a minimal example, and enabling `DEBUG` in `InferenceSettings` to print out subtyping constraints as they are determined, and then find which constraint is being incorrectly added and why.